### PR TITLE
FEATURE: Show when a badge has been granted for a post

### PR DESCRIPTION
**Pull request by pento**
(New PR copied from #29362, moving the source branch to the `discourse` repo).

## ✨ What's This?

Based on [this feature request](https://meta.discourse.org/t/wishlist-visual-cue-for-user-got-a-badge-for-this-post/282923/1), this PR adds a visual cue to the post header when the user was granted a badge for that post.

You can customise the appearance of these badges (or hide some of them entirely!) by adding custom theme CSS. For example:

```css
.topic-meta-data .user-badge-button-first-onebox {
  display: none;
}

.topic-meta-data .user-badge-button-first-onebox {
  animation: blinker 1s step-start infinite;
}

@keyframes blinker {
  50% {
    opacity: 0;
  }
}
```

There's also a global option to turn this feature off, and each individual badge has an option to decide if it should be included in the post header or not.

### Per-Badge Setting

https://github.com/user-attachments/assets/a4ec7972-aa88-4433-a835-cac284c9e6af

## 👑 Testing

Badges should only ever show on a post when all of these conditions are true:
- The badge was granted to the post author.
- The badge was granted for that post.
- The badge is enabled.
- The badge can be shown on the public badge page.
- The badge allows showing the post that granted this badge on the badge page.
- The badge allows being displayed in the post header.

Most of the testing should be ensuring that these conditions can't be avoided, and that there are no other conditions to add.

It'd also be helpful to test badge customisation with theme CSS, to ensure it can't be accidentally made to look weird. Looking weird on purpose is totally fine, though! 😄

### CODE REVIEW

File: app/assets/javascripts/discourse/app/form-kit/components/fk/control/checkbox.gjs [hunk 1]
```diff
@@ -8,8 +8,8 @@ export default class FKControlCheckbox extends Component {
   static controlType = "checkbox";
 
   @action
-  handleInput() {
-    this.args.field.set(!this.args.value);
+  handleInput(event) {
+    this.args.field.set(event.target.checked);
```
Comment (pento) [inline app/assets/javascripts/discourse/app/form-kit/components/fk/control/checkbox.gjs:12]: This will need to be broken out into its own PR, just keeping it here for ease of iteration for now.

File: app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs [hunk 2]
```diff
@@ -280,10 +285,31 @@
...
         as |field|
       >
         <field.Checkbox />
       </group.Field>
+
+      <group.Field
+        @title={{i18n "admin.badges.post_header"}}
+        @showTitle={{false}}
+        @name="post_header"
+        @disabled={{this.disableBadgeOnPosts}}
```
Comment (pento) [inline app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs:298]: It seems like this is only run once, on initial render. It isn't being re-run when the other checkboxes are changed.

File: app/assets/javascripts/discourse/app/form-kit/components/fk/control/checkbox.gjs (hunk 1 above)
Comment (OsamaSayegh) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/control/checkbox.gjs:12]: Can you clarify the problem that this fixes? Your change looks good to me, but I can't see anything wrong in old code either. Unless it's somehow possible to trigger an `input` event twice without changing the state of the checkbox?

File: app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs (hunk 2 above)
Comment (jjaffeux) [reply app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs:298]: Ah yes it's on my list to investigate, will look thanks for reminder :+1:

Comment (jjaffeux) [reply app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs:298]: Fixed in: https://github.com/discourse/discourse/pull/29738

File: app/assets/javascripts/discourse/app/form-kit/components/fk/control/checkbox.gjs (hunk 1 above)
Comment (pento) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/control/checkbox.gjs:12]: The issue here is that `this.args.value` always seems to be the initial value, rather than reflecting the current checked stated of the checkbox, when means that the `onSet` handler will only ever receive the same value, rather than having the value change if the checkbox is checked/unchecked multiple times.

File: app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs (hunk 2 above)
Comment (pento) [reply app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs:298]: That fixed it here, thanks @jjaffeux!

File: app/assets/javascripts/discourse/app/form-kit/components/fk/control/checkbox.gjs (hunk 1 above)
Comment (pento) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/control/checkbox.gjs]: Okay, I figured out the problem. Setting `onSet` causes `set` to not be called on the data model, which caused the behaviour I was observing.

That said, I'd expect listening to a psuedo-event (ie, `onSet`) to not have side effects, so perhaps that behaviour should be changed? That's a much bigger change, though.

Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/form-kit/components/fk/control/checkbox.gjs]: I have pushed what I would expect in this situation: https://github.com/discourse/discourse/pull/29696/commits/539aeea4c04f044f39ec55d4cc9676dd81f45772

File: app/assets/javascripts/discourse/app/components/badge-button.gjs [hunk 3]
```diff
@@ -10,6 +10,11 @@ export default class BadgeButton extends Component {
     }
   }
 
+  get showName() {
+    // If the showName argument is not set, default to true
+    return this.args.showName !== false;
```
Comment (CvX) [inline app/assets/javascripts/discourse/app/components/badge-button.gjs:15]: making the code self explanatory

```suggestion
    return this.args.showName ?? true;
```

File: app/assets/javascripts/discourse/app/lib/transform-post.js [hunk 4]
```diff
@@ -37,6 +38,9 @@ export function transformBasicPost(post) {
     user_id: post.user_id,
     usernameUrl: userPath(post.username),
     username: post.username,
+    badgesGranted:
+      post.badges_granted &&
+      post.badges_granted.map((badge) => Badge.createFromJson(badge)[0]),
```
Comment (CvX) [inline app/assets/javascripts/discourse/app/lib/transform-post.js:43]: null operator for a bit of code golf

```suggestion
      post.badges_granted?.map((badge) => Badge.createFromJson(badge)[0]),
```

File: config/locales/client.en.yml [hunk 5]
```diff
@@ -7151,9 +7153,13 @@ en:
         no_user_badges: "%{name} has not been granted any badges."
         no_badges: There are no badges that can be granted.
         none_selected: "Select a badge to get started"
+        usage_heading: Usage
         allow_title: Allow badge to be used as a title
         multiple_grant: Can be granted multiple times
+        visibility_heading: Visibility
         listable: Show badge on the public badges page
+        post_header: Show badge on the post it was granted for
+        post_header_disabled: Requires both ‘Show badge on the public badges page’ and ‘Show post granting badge on badge page’ to be enabled.
```
Comment (CvX) [inline config/locales/client.en.yml:7162]: for consistency (there's only one instance of `‘` quote in the file currently)

```suggestion
        post_header_disabled: Requires both 'Show badge on the public badges page' and 'Show post granting badge on badge page' to be enabled.
```

File: app/assets/javascripts/admin/addon/controllers/admin-badges/show.js [hunk 6]
```diff
@@ -80,6 +79,17 @@ export default class AdminBadgesShowController extends Controller {
     return this.model.system;
   }
 
+  @action
+  postHeaderDescription(data = {}) {
```
Comment (CvX) [inline app/assets/javascripts/admin/addon/controllers/admin-badges/show.js:83]: are those two actions ever called with a null-ish arg?

Comment (jjaffeux) [reply app/assets/javascripts/admin/addon/controllers/admin-badges/show.js:83]: Yes, it was a surprise to me too. I need to look how that can happen.

Comment (jjaffeux) [reply app/assets/javascripts/admin/addon/controllers/admin-badges/show.js]: I have been checking this enough... and couldn't repro the empty data anymore, it was probably related to another error. I pushed a change to remove the `= {}`

File: app/assets/javascripts/discourse/tests/integration/components/badge-button-test.js [hunk 7]
```diff
@@ -77,4 +77,22 @@ module("Integration | Component | badge-button", function (hooks) {
 
     assert.dom(".user-badge.foo").exists();
   });
+
+  test("showName hides the name", async function (assert) {
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/tests/integration/components/badge-button-test.js:81]:
```suggestion
  test("setting showName to false hides the name", async function (assert) {
```

Small suggestion cause we can also set `showName` to `true`.

File: app/models/post.rb [hunk 8]
```diff
@@ -217,6 +222,10 @@ def limit_posts_per_day
     end
   end
 
+  def badges_granted
+    user_badges.select { |user_badge| user_badge.user_id == user_id }
```
Comment (tgxworld) [inline app/models/post.rb:226]:
```suggestion
    user_badges.where("user_badges.user_id = ?", user_id)
```

I believe we can do the filtering directly in the database to avoid over fetching from the database.

File: app/serializers/post_serializer.rb [hunk 9]
```diff
@@ -223,6 +224,14 @@ def flair_group_id
     object.user&.flair_group_id
   end
 
+  def badges_granted
+    return [] unless SiteSetting.enable_badges && SiteSetting.show_badges_in_post_header
+
+    object.badges_granted.map do |user_badge|
```
Comment (tgxworld) [inline app/serializers/post_serializer.rb:230]: This will result in the N+1 queries problem since the SQL query will be executed once per post that is being serialized. From what I can tell, the existing pattern seems to be to preload within `TopicView` and access the preloaded data via the `@topic_view` instance variable in the serializer.

File: app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs [hunk 10]
```diff
@@ -284,6 +289,20 @@
       >
         <field.Checkbox />
       </group.Field>
+
+      <group.Field
```
Comment (tgxworld) [inline app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs:293]: `this.siteSettings.show_badges_in_post_header` is `false` and I'm wondering if we should just hide this field completely when it is so?

File: db/migrate/20241030000131_add_post_header_to_badge.rb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+class AddPostHeaderToBadge < ActiveRecord::Migration[7.1]
+  def change
+    add_column :badges, :post_header, :boolean, default: false, null: false
```
Comment (tgxworld) [inline db/migrate/20241030000131_add_post_header_to_badge.rb:4]:
```suggestion
    add_column :badges, :show_in_post_header, :boolean, default: false, null: false
```

Maybe something like the suggestion above for the column name? It is more verbose but I find it much clearer in indicating what the column is used for.

File: app/serializers/post_serializer.rb (hunk 9 above)
Comment (tgxworld) [reply app/serializers/post_serializer.rb:230]: My bad, I see that you have already eager loaded this in `TopicView`.

File: app/models/post.rb [hunk 12]
```diff
@@ -67,6 +67,11 @@ class Post < ActiveRecord::Base
 
   has_many :user_actions, foreign_key: :target_post_id
 
+  has_many :user_badges,
+           -> { for_post_header_badges },
+           foreign_key: :post_id,
+           class_name: "UserBadge"
```
Comment (tgxworld) [inline app/models/post.rb:73]:
```suggestion
  has_many :user_badges,
           ->(post) { for_post_header_badges.where("badges.user_id = ?", post.user_id) },
           foreign_key: :post_id,
           class_name: "UserBadge"
```

File: spec/serializers/post_serializer_spec.rb [hunk 13]
```diff
@@ -438,6 +438,71 @@ def json_for_user(user)
...
+        user: user,
+        granted_by: Discourse.system_user,
+        granted_at: Time.now,
+        post_id: post.id,
+      )
+    end
+
+    let(:serializer) { described_class.new(post, scope: Guardian.new(user), root: false) }
+
+    it "doesn't include badges when disabled" do
```
Comment (tgxworld) [inline spec/serializers/post_serializer_spec.rb:481]:
```suggestion
    it "doesn't include badges when `enable_badges` site setting is disabled" do
```

I find `when disabled` alittle vague because it doesn't specify what is actually being disabled.

File: spec/serializers/post_serializer_spec.rb [hunk 14]
```diff
@@ -438,6 +438,71 @@ def json_for_user(user)
...
+      SiteSetting.enable_badges = false
+      expect(serializer.as_json[:badges_granted]).to eq([])
+    end
+
+    it "doesn't include badges when display is disabled" do
```
Comment (tgxworld) [inline spec/serializers/post_serializer_spec.rb:486]:
```suggestion
    it "doesn't include badges when `show_badges_in_post_header` site setting is disabled" do
```

File: spec/serializers/post_serializer_spec.rb [hunk 15]
```diff
@@ -438,6 +438,71 @@ def json_for_user(user)
...
+      SiteSetting.enable_badges = true
+      SiteSetting.show_badges_in_post_header = false
+      expect(serializer.as_json[:badges_granted]).to eq([])
+    end
+
+    it "includes badges when enabled" do
```
Comment (tgxworld) [inline spec/serializers/post_serializer_spec.rb:492]:
```suggestion
    it "includes badges when `enable_badges` and `show_badges_in_post_header` site settings are enabled" do
```

**Review by tgxworld — COMMENTED** [review]
@pento I think it'll be good to have some happy-path end-to-end system tests to ensure that all the changes work together as expected.

File: app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs (hunk 10 above)
Comment (pento) [reply app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs:293]: That'd make this option very difficult to discover organically, you'd have to know how to get to it first before you could enable it.

Comment (tgxworld) [reply app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs:293]: This is fair, thank you for clarifying 👍

**Comment by pento** [comment]
@tgxworld: Thank you for all the feedback! I've gone through and addressed it all, I'd appreciate you having another look. 🙂

File: lib/topic_view.rb [hunk 16]
```diff
@@ -135,6 +135,8 @@ def initialize(topic_or_topic_id, user = nil, options = {})
     @filtered_posts = apply_default_scope(@filtered_posts)
     filter_posts(options)
 
+    @post_user_badges = UserBadge.for_post_header_badges(@posts) if @posts
```
Comment (tgxworld) [inline lib/topic_view.rb:138]: I think we need to consider the `SiteSetting.enable_badges && SiteSetting.show_badges_in_post_header` site settings here. We don't want to be eager loading when we don't need to. I think what we can do something like this for the `post_user_badges` method?

```
def post_user_badeges
  @post_user_badges ||= begin
    UserBadge.for_post_header_badges(@posts).reduce({}) do |hash, user_badge|
      hash[user_badge.post_id] ||= []
      hash[user_badge.post_id] << user_badge
      hash
    end
  end
end
```

We memoize the results so that the query is only triggered once even if the method is called multiple times from different instance of `PostSerializer`. We also return a hash instead so that the look up from the `PostSerializer` does not involve looping through all the user badges for all the posts each time.

File: spec/serializers/post_serializer_spec.rb [hunk 17]
```diff
@@ -438,6 +438,197 @@ def json_for_user(user)
...
+        SiteSetting.enable_badges = true
+        SiteSetting.show_badges_in_post_header = true
+      end
+
+      it "includes badges that were granted for this user on this post" do
+        json = serializer.as_json
+
+        expect(json[:badges_granted].length).to eq(2)
+        expect(json[:badges_granted].map { |b| b[:badges][0][:id] }).to include(ub1.badge_id)
+        expect(json[:badges_granted].map { |b| b[:badges][0][:id] }).to include(ub2.badge_id)
```
Comment (tgxworld) [inline spec/serializers/post_serializer_spec.rb:594]:
```suggestion
        expect(json[:badges_granted].map { |b| b[:badges][0][:id] }).to contain_exactly(ub1.badge_id, ub2.badge_id)
```

File: spec/serializers/post_serializer_spec.rb [hunk 18]
```diff
@@ -438,6 +438,197 @@ def json_for_user(user)
...
+
+        expect(json[:badges_granted].map { |b| b[:basic_user_badge][:id] }).to include(ub1.id)
+        expect(json[:badges_granted].map { |b| b[:basic_user_badge][:id] }).to include(ub2.id)
```
Comment (tgxworld) [inline spec/serializers/post_serializer_spec.rb:597]:
```suggestion
        expect(json[:badges_granted].map { |b| b[:basic_user_badge][:id] }).to contain_exactly(ub1.id, ub2.id)
```

File: spec/serializers/web_hook_post_serializer_spec.rb [hunk 19]
```diff
@@ -33,6 +33,7 @@ def serialized_for_user(u)
       :primary_group_name,
       :flair_name,
       :flair_group_id,
+      :badges_granted,
```
Comment (tgxworld) [inline spec/serializers/web_hook_post_serializer_spec.rb:36]: I don't think we need to include the `badges_granted` attribute in the `web_hook_post_serializer` as it isn't very useful for webhooks. You can exclude it by adding the key to [this array](https://github.com/discourse/discourse/blob/72132c35fb3f90b0323c6ce735e4e7bd4e783100/app/serializers/web_hook_post_serializer.rb#L18-L33).

**Review by tgxworld — APPROVED** [review]
Left a few more comments but I think the PR is looking good 👍
